### PR TITLE
Detect circularities printing recursive conditional types

### PR DESCRIPTION
**Pull request by andrewbranch**
This fixes the simplified cases of #43529, but ~~the full repro from the playground there still OOMs~~. The total print depth (measured by incrementing and decrementing a counter going into and out of `typeToTypeNodeHelper`) seems to max out in the 30s, so I’m not sure how it’s getting into an infinite(?) loop without recurring deeper and deeper.

Also, even where this change successfully prevents a crash, I think we could do better somehow. I experimented with some changes to detect where a type alias could be used to stop recursively printing the true or false branch of a conditional, but I think it didn’t correctly account for the visibility of that type alias.

Putting this up as is and asking @ahejlsberg or @weswigham for some help 🙏. (The bug is not a 4.2 or 4.3 regression so this may be lower priority than other things on your plate for 4.3.)

### CODE REVIEW

**Comment by andrewbranch** [comment]
With https://github.com/microsoft/TypeScript/pull/43973, these tests no longer OOM. However, `recursiveConditionalCrash3.types` is alarmingly bad. GitHub won’t even show it, so here’s a video summary:

https://user-images.githubusercontent.com/3277153/120039904-83bdac00-bfba-11eb-94fa-4043959ea60a.mp4

**Review by weswigham — APPROVED** [review]
It's possible the printback in the super long baseline is only allowed to be as such because cached node trees don't contribute to the estimated length of the output string. We could probably trim down how much we're willing to print back a bunch by saving that information.

**Comment by andrewbranch** [comment]
@weswigham is that not what the `addedLength` property on `cachedResult` is doing? I was just going to try out your suggestion, but what I had in my mind seems to be already done, yet the baseline output remains outrageously long.

**Comment by weswigham** [comment]
Heh, well, the no-truncation limit is something like a million characters...

File: tests/baselines/reference/arrayFakeFlatNoCrashInferenceDeclarations.types [hunk 1]
```diff
@@ -7,7 +7,7 @@ type BadFlatArray<Arr, Depth extends number> = {obj: {
...
..., 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][Depth]]]]] extends -1 ? "done" : "recur"] : InnerArr; }[[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][Depth]]]] extends -1 ? "done" : "recur"] : InnerArr; }[[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][Depth]]] extends -1 ? "done" : "recur"] : InnerArr; }[[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][Depth]] extends -1 ? "done" : "recur"] : InnerArr; }[[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20][Depth] extends -1 ? "done" : "recur"] : Arr
```
Comment (andrewbranch) [inline tests/baselines/reference/arrayFakeFlatNoCrashInferenceDeclarations.types:10]: Uh, is this change ok @weswigham?

Comment (weswigham) [reply tests/baselines/reference/arrayFakeFlatNoCrashInferenceDeclarations.types:10]: I'm... Not even sure what's different at a glance

Comment (andrewbranch) [reply tests/baselines/reference/arrayFakeFlatNoCrashInferenceDeclarations.types:10]: Screen Shot 2021-07-06 at 3 44 07 PM

Comment (weswigham) [reply tests/baselines/reference/arrayFakeFlatNoCrashInferenceDeclarations.types:10]: Uhh... Yeah, it's probably fine - there's an `any` due to truncation in the removed type text, so all we did was truncate more, which is generally fine.

**Pull request merged**